src/bp/nlu-core/training-pipeline.ts

```diff
@@ -537,11 +556,11 @@ export const Trainer = async (
   let step = await PreprocessInput(input, tools)
   reportProgress() // 10%
 
-  step = await TfidfTokens(step)
+  step = await TfidfTokens(step, tools)
```

Comment:
Just a personal opinion, feel free to disagree:

Usually, In the context of a multi-step sequential algorithm, I prefer to add logging statements in the "main" algorithm rather than inside the individual sub-functions. e.g.:
```
tools.logger?.debug(`[${input.trainId}] Computing TFIDF`)
step = await TfidfTokens(step, tools)
tools.logger?.debug(`[${input.trainId}] Done computing TFIDF`)
step = ClusterTokens(step, tools)
tools.logger?.debug(`[${input.trainId}] Extracting entities`)
step = await ExtractEntities(step, tools)
tools.logger?.debug(`[${input.trainId}] Done extracting entities`)
...
```

This keeps the individual functions smaller, allows reusing the functions without logging and concentrates logging in a single place.

Logging statements inside the sub-functions should add details about the internal workings of the sub-functions, e.g.:
```
export const TfidfTokens = async (input: TrainStep, tools: Tools): Promise<TrainStep> => {

  const tfidfInput = input.intents.reduce(
    (tfidfInput, intent) => ({
      ...tfidfInput,
      [intent.name]: _.flatMapDeep(intent.utterances.map(u => u.tokens.map(t => t.toString({ lowerCase: true }))))
    }),
    {} as _.Dictionary<string[]>
  )

  tools.logger?.debug(`[${input.trainId}] TFIDF: ${tfidfInput}`)

  const { __avg__: avg_tfidf } = tfidf(tfidfInput)
  const copy = { ...input, tfIdf: avg_tfidf }
  copy.intents.forEach(x => x.utterances.forEach(u => u.setGlobalTfidf(avg_tfidf)))

  tools.logger?.debug(`[${input.trainId}] TFIDF: ${copy}`)

  return copy
}
```

Reply:
yep, good point here.